src/wallet/wallet2.cpp

```diff
@@ -606,11 +627,48 @@ static uint64_t decodeRct(const rct::rctSig & rv, const crypto::public_key &pub,
   }
 }
 //----------------------------------------------------------------------------------------------------
-bool wallet2::wallet_generate_key_image_helper(const cryptonote::account_keys& ack, const crypto::public_key& tx_public_key, size_t real_output_index, cryptonote::keypair& in_ephemeral, crypto::key_image& ki)
+static uint64_t decodeRct_onetime(const rct::rctSig & rv, const crypto::public_key pub, const crypto::secret_key &sec, bool received_onetime, const crypto::secret_key* onetime_c, unsigned int i, rct::key & mask)
 {
-  if (!cryptonote::generate_key_image_helper(ack, tx_public_key, real_output_index, in_ephemeral, ki))
-    return false;
-  return true;
+  if (received_onetime)
+  {
+    return decodeRct(rv, pub, *onetime_c, i, mask);
+  }
+  else
+  {
+    return decodeRct(rv, pub, sec, i, mask);
+  }
+}
+//----------------------------------------------------------------------------------------------------
+struct payment_id_info {
+  tx_extra_nonce extra_nonce;
+  bool has_payment_id32 = false;
+  bool has_payment_id8  = false;
+  crypto::hash  payment_id32 = null_hash ;
+  crypto::hash8 payment_id8  = null_hash8;
+  crypto::secret_key* onetime_c = nullptr;
+  crypto::secret_key  onetime_c_real;
+};
+static void get_payment_id_info(const std::vector<tx_extra_field>& tx_extra_fields, const crypto::secret_key& viewkey, payment_id_info& pID_info) {
+  if (find_tx_extra_field_by_type(tx_extra_fields, pID_info.extra_nonce))
+  {
+    pID_info.has_payment_id32 = get_payment_id_from_tx_extra_nonce          (pID_info.extra_nonce.nonce, pID_info.payment_id32);
+    pID_info.has_payment_id8  = get_encrypted_payment_id_from_tx_extra_nonce(pID_info.extra_nonce.nonce, pID_info.payment_id8 );
+    // if 8-byte payment ID is found, copy it as-is to 32-byte payment ID regardless of whether it's encrypted or not
+    if (pID_info.has_payment_id8)
+    {
+      memcpy(&pID_info.payment_id32, &pID_info.payment_id8, 8);
+      memset(pID_info.payment_id32.data + 8, 0, 24);
+    }
+    // try to get h for onetime address scheme
+    static_assert(sizeof(viewkey) == HASH_SIZE && sizeof(pID_info.payment_id32) == HASH_SIZE, "Invalid structure size");
+    char data[2 * HASH_SIZE];
+    memcpy(data            , &viewkey              , HASH_SIZE);
+    memcpy(data + HASH_SIZE, &pID_info.payment_id32, HASH_SIZE);
+    crypto::hash_to_scalar(data, 2 * HASH_SIZE, reinterpret_cast<crypto::ec_scalar&>(pID_info.onetime_c_real));
+    // check if the following equation holds: (H(viewkey, pID) - pID) mod 256 == 0
+    if (pID_info.payment_id32.data[0] == pID_info.onetime_c_real.data[0])
+      pID_info.onetime_c = &pID_info.onetime_c_real;
```

Comment:
Well like above with `is_received_info`, all fields of `pID_info` are default initialized as:
```
struct payment_id_info {
  tx_extra_nonce extra_nonce;
  bool has_payment_id32 = false;
  bool has_payment_id8  = false;
  crypto::hash  payment_id32 = null_hash ;
  crypto::hash8 payment_id8  = null_hash8;
  crypto::secret_key* onetime_c = nullptr;
  crypto::secret_key  onetime_c_real;
};
```
Should I stop this and use your style?